gcc/rust/rust-diagnostics.h

```diff
@@ -48,1047 +48,33 @@
-TABLE_TO_MAP(E0698),
-TABLE_TO_MAP(E0699),
-TABLE_TO_MAP(E0700),
-TABLE_TO_MAP(E0701),
-TABLE_TO_MAP(E0703),
-TABLE_TO_MAP(E0704),
-TABLE_TO_MAP(E0705),
-TABLE_TO_MAP(E0706),
-TABLE_TO_MAP(E0708),
-TABLE_TO_MAP(E0710),
-TABLE_TO_MAP(E0711),
-TABLE_TO_MAP(E0712),
-TABLE_TO_MAP(E0713),
-TABLE_TO_MAP(E0714),
-TABLE_TO_MAP(E0715),
-TABLE_TO_MAP(E0716),
-TABLE_TO_MAP(E0717),
-TABLE_TO_MAP(E0718),
-TABLE_TO_MAP(E0719),
-TABLE_TO_MAP(E0720),
-TABLE_TO_MAP(E0722),
-TABLE_TO_MAP(E0724),
-TABLE_TO_MAP(E0725),
-TABLE_TO_MAP(E0726),
-TABLE_TO_MAP(E0727),
-TABLE_TO_MAP(E0728),
-TABLE_TO_MAP(E0729),
-TABLE_TO_MAP(E0730),
-TABLE_TO_MAP(E0731),
-TABLE_TO_MAP(E0732),
-TABLE_TO_MAP(E0733),
-TABLE_TO_MAP(E0734),
-TABLE_TO_MAP(E0735),
-TABLE_TO_MAP(E0736),
-TABLE_TO_MAP(E0737),
-TABLE_TO_MAP(E0739),
-TABLE_TO_MAP(E0740),
-TABLE_TO_MAP(E0741),
-TABLE_TO_MAP(E0742),
-TABLE_TO_MAP(E0743),
-TABLE_TO_MAP(E0744),
-TABLE_TO_MAP(E0745),
-TABLE_TO_MAP(E0746),
-TABLE_TO_MAP(E0747),
-TABLE_TO_MAP(E0748),
-TABLE_TO_MAP(E0749),
-TABLE_TO_MAP(E0750),
-TABLE_TO_MAP(E0751),
-TABLE_TO_MAP(E0752),
-TABLE_TO_MAP(E0753),
-TABLE_TO_MAP(E0754),
-TABLE_TO_MAP(E0755),
-TABLE_TO_MAP(E0756),
-TABLE_TO_MAP(E0757),
-TABLE_TO_MAP(E0758),
-TABLE_TO_MAP(E0759),
-TABLE_TO_MAP(E0760),
-TABLE_TO_MAP(E0761),
-TABLE_TO_MAP(E0762),
-TABLE_TO_MAP(E0763),
-TABLE_TO_MAP(E0764),
-TABLE_TO_MAP(E0765),
-TABLE_TO_MAP(E0766),
-TABLE_TO_MAP(E0767),
-TABLE_TO_MAP(E0768),
-TABLE_TO_MAP(E0769),
-TABLE_TO_MAP(E0770),
-TABLE_TO_MAP(E0771),
-TABLE_TO_MAP(E0772),
-TABLE_TO_MAP(E0773),
-TABLE_TO_MAP(E0774),
-TABLE_TO_MAP(E0775),
-TABLE_TO_MAP(E0776),
-TABLE_TO_MAP(E0777),
-TABLE_TO_MAP(E0778),
-TABLE_TO_MAP(E0779),
-TABLE_TO_MAP(E0780),
-TABLE_TO_MAP(E0781),
-TABLE_TO_MAP(E0782),
-TABLE_TO_MAP(E0783),
-TABLE_TO_MAP(E0784),
-TABLE_TO_MAP(E0785),
-TABLE_TO_MAP(E0786),
-TABLE_TO_MAP(E0787),
-TABLE_TO_MAP(E0788),
-TABLE_TO_MAP(E0789),
-TABLE_TO_MAP(E0790),
-TABLE_TO_MAP(E0791),
-TABLE_TO_MAP(E0792),
-TABLE_TO_MAP(E0793),
-TABLE_TO_MAP(E0794),
+// We want E0005 to be mapped to the value `5` - this way, we can easily format
+// it in `make_description`. We also want to keep the value "5" only once when
+// defining the error code in rust_error_codes.def, so not have ERROR(E0005, 5)
+// as that is error prone. If we just use `0005` as the discriminant for the
+// `E0005` enum variant, then we are actually creating octal values (!) as `0`
+// is the C/C++ octal prefix. So this does not work for `E0009` for example,
+// since 0009 is not a valid octal literal.
+// We can circumvent this by doing a little bit of constant folding in the
+// discriminant expression. So for ERROR(E0009), this macro expands to the
+// following variant:
+//
+// E0009 = (10009 - 10000)
+//
+// which gets folded to the result of the substraction... 9. A valid decimal
+// literal which corresponds to E0009.
+#define ERROR(N) E##N = (1##N - 10000)
+enum class ErrorCode : unsigned int
+{
+#include "rust_error_codes.def"
 };
+#undef ERROR
```

Comment:
Sorry for late review, LGTM , that's a great logic  for converting `enum` to string by assigning them the same non-negative numbers. 

I am a bit sad for one thing that we no longer get the code completion for enum class, but that would be fine :)